include/nbl/builtin/hlsl/matrix_utils/transformation_matrix_utils.hlsl

```diff
@@ -1,15 +1,45 @@
-#ifndef _NBL_BUILTIN_HLSL_TRANSFORMATION_MATRIX_UTILS_INCLUDED_
-#define _NBL_BUILTIN_HLSL_TRANSFORMATION_MATRIX_UTILS_INCLUDED_
+#ifndef _NBL_BUILTIN_HLSL_MATRIX_UTILS_TRANSFORMATION_MATRIX_UTILS_INCLUDED_
+#define _NBL_BUILTIN_HLSL_MATRIX_UTILS_TRANSFORMATION_MATRIX_UTILS_INCLUDED_
 
 #include <nbl/builtin/hlsl/cpp_compat.hlsl>
+#include <nbl/builtin/hlsl/math/quaternion/quaternion.hlsl>
+// TODO: remove this header when deleting vectorSIMDf.hlsl
+#include <nbl/core/math/glslFunctions.h>
+#include "vectorSIMD.h"
 
 namespace nbl
 {
 namespace hlsl
 {
 
+// TODO: if `IdentityFloat32_t3x4` and `IdentityFloat32_t3x4` constexprs are ok, then I can expand them into templated struct, not doing it untill the concept is approved
+//template<typename T, uint32_t N, uint32_t M>
+//struct IdentityMatrix
```

Comment:
add

```c++
template<typename T>
concept MatrixType = requires 
{
  typename T::Base::row_type;
  typename T::Base::col_type;
  typename T::Base::value_type;
} && std::is_base_of_v<glm::mat<T::Base::row_type::length(), T::Base::col_type::length(), typename T::Base::value_type>, typename T::Base>;
```

to `matrix.hlsl` just after [definition](<https://github.com/Devsh-Graphics-Programming/Nabla/blob/rm_core_mat_vec/include/nbl/builtin/hlsl/cpp_compat/matrix.hlsl#L49 >) (but ask @devshgraphicsprogramming if it should be named `MatrixType`)

then create `diagonal` utility

```c++
template<MatrixType Matrix>
constexpr inline Matrix diagonal(float scalar)
{
  return Matrix(scalar);
}
```

you can also template the scalar actually and make it `value_type` of the `Matrix`, now note that

- `matrix<T, N, M>(scalar)` creates a [diagonal matrix](<https://en.wikipedia.org/wiki/Diagonal_matrix>) with `scalar` put on the `diagonal`
- if your `N == M` then `matrix<T, N, M>(1.f)` is the identity matrix you need

so following `diagonal` create `identity` function (not struct!), you should be able to 

```c++
using any_square_matrix_t = float32_t3x3;
constexpr auto bruh = hlsl::identity<any_square_matrix_t>()
``` 

have constraints on the template parameter, let's allow only for square matrices

Reply:
but this is not C++ only file

Reply:
so we can't use concepts..

Reply:
> so we can't use concepts..

hehe, you can! https://github.com/Devsh-Graphics-Programming/Nabla/commit/492a0ad804b65e656c8a7f29fbd5b16d7d58dc92

Reply:
```cpp
  typename T::Base::row_type;
  typename T::Base::col_type;
  typename T::Base::value_type;
  ```
  this is specific to GLM and won't pass in plain HLSL